Introduce UniquelyOwned and InlineWeakPtr and use them in RenderLayer

geoffreygaren opened the pull request:
#### c7f7a508727420ada216e981d0c7d6fd99a199f7
<pre>
Introduce UniquelyOwned and InlineWeakPtr and use them in RenderLayer
<a href="https://bugs.webkit.org/show_bug.cgi?id=304082">https://bugs.webkit.org/show_bug.cgi?id=304082</a>
&lt;<a href="https://rdar.apple.com/problem/166403158">rdar://problem/166403158</a>&gt;

Reviewed by Ryosuke Niwa.

Example usage:

class T : public UniquelyOwned&lt;T&gt; {
public:
    UniquelyOwnedPtr&lt;T&gt; create() { return adoptUniquelyOwned(new T); }

private:
    T() { }
};

class Owner {
    UniquelyOwnedPtr&lt;T&gt; m_t;
};

class X {
    // Data members must be weak
    InlineWeakPtr&lt;T&gt; m_t1;
    InlineWeakRef&lt;T&gt; m_t2;
};

void X::f()
{
    // Locals must be Checked
    if (CheckedPtr checkedT1 = m_t1)
        checkedT1-&gt;g();
    ChekckedRef { m_t2 }-&gt;g();

    // CheckedPtrs in the heap are also OK, but only if they&apos;re temporary,
    // and out of scope before ~T().
    Vector&lt;CheckedPtr&lt;T&gt;&gt; tmpVector;
    HashSet&lt;CheckedPtr&lt;T&gt;&gt; tmpHashSet;
}

This patch is neutral on Speedometer, MotionMark, and Membuster.

Chris tried diligently to adopt existing smart pointers in RenderLayer data
members, but it didn&apos;t work out:

    * SingleThreadWeakPtr is a performance regression

    * CheckedPtr crashes because our pointers outlive their pointees

    * Clearing pointers eagerly to prevent CheckedPtr crashes is a performance
    regression

Solution: InlineWeakPtr. InlineWeakPtr is a counter directly inside the object.

InlineWeakPtr is always faster than SingleThreadWeakPtr because it doesn&apos;t
require allocation or double-indirection.

InlineWeakPtr is sometimes smaller and sometimes larger than
SingleThreadWeakPtr:
    * It&apos;s 24 bytes smaller when the pointee is live
    * It&apos;s equal in size when the pointee and all pointers are dead
    * It&apos;s sizeof(T) - 16 bytes larger when the pointee is dead but a
    pointer is still live.

So InlineWeakPtr is always faster, and requires some thoughtfulness about
pointee lifetime and size.

InlineWeakPtr depends internally on reference counting
(RefCountedWithInlineWeakPtr), but we want to maintain the clarity that
RenderLayer is a single-owner object. So this patch also introduces UniquelyOwned
and UniquelyOwnedPtr, which indicate and enforce single ownership explicitly.

UniquelyOwned also improves the safety of CheckedPtr by eliminating
    * the window of time after scribbling and before crashing during which
    pointers borrowed before scribbling might still UAF
    * the whacky macro that must be included in all subclasses

I designed UniquelyOwned and RefCountedWithInlineWeakPtr to be useable in many
more places, including thread-safe places -- but this patch doesn&apos;t take those
steps yet.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/InlineWeakPtr.h: InlineWeakPtr is basically a copy of
RefPtr&lt;T&gt;, except that it calls weakRef() / weakDeref() instead of ref() / deref
(), and it checks for a non-zero refCount() before returning T*.

* Source/WTF/wtf/InlineWeakRef.h: Just like WeakRef&lt;T&gt;, for use in cases where
we want to verify memory safety but writing code to handle null would be absurd.

* Source/WTF/wtf/RefCountedWithInlineWeakPtr.h: The key insight of
RefCountedWithInlineWeakPtr is that the refcount gives us total control over
destruction, so we get a much tighter interface for checking validity of weak
pointers and CheckedPtrs.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/RefCntblBaseVirtualDtorExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
Updated expectations to copy RenderLayer CheckedPtr skips to RefCounted skips,
since UniquelyOwned is internally backed by refcounting.

* Source/WebCore/rendering/RenderLayer.h: Replace CanMakeSingleThreadWeakPtr&lt;T&gt;,
CanMakeCheckedPtr&lt;T&gt;, and WTF_OVERRIDE_DELETE_FOR_CHECKED_PTR with
UniquelyOwned&lt;T&gt;. Yay!

* Source/WebCore/rendering/RenderMarquee.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceContainer.cpp:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceContainer.h:
Adopt InlineWeakPtr now that RenderLayer requires it.

* Tools/TestWebKitAPI/Tests/WTF/CheckedPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/HashMap.cpp:
* Tools/TestWebKitAPI/Tests/WTF/HashSet.cpp:
* Tools/TestWebKitAPI/Tests/WTF/ListHashSet.cpp: Some API tests for new
CheckedPtr and InlineWeakPtr usage.

Canonical link: <a href="https://commits.webkit.org/304729@main">https://commits.webkit.org/304729@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be36eb65b3200d2d73af8718ed1fc7712845f0da

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136390 "8 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8747 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47670 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144102 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8e7238bc-49ad-4554-9ffe-9cbdc6875a05) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138262 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9430 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8591 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104291 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/63653a33-c20e-4b1d-aae9-0c4637a984c8) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139335 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6871 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122204 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85127 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/774e6827-0e49-414c-83e9-003f773e4ac6) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6518 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4175 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4693 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128347 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115812 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40410 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146846 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134874 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8429 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40979 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112631 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8446 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7077 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112978 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6451 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118514 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62416 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21024 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8477 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36566 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167653 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8195 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72036 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43732 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8417 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8269 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WTF/wtf/SinglyOwned.h:
```diff
@@ -0,0 +1,87 @@
+//     * In debug builds, the CheckedPtr ASSERTs
+//     * In release builds, the CheckedPtr keeps the pointed-to object alive.
+template<typename T>
+class SinglyOwned : private RefCountedWithInlineWeakPtr<T> {
```
Maybe we should call this `CanMakeUniqueCheckedWeakPtr`?
Or if we want a shorter name, I'd call this `UniquelyOwned` to use a more assimilating name.

geoffreygaren replied:
`CanMakeUniqueCheckedWeakPtr` is too many disjoint words for my taste. Also we can't make a `Unique...Ptr`.

I liked `SinglyOwned` because in conversation we usually talk about something having "single ownership".

I'll start a Slack thread...

rniwa commented on Source/WTF/wtf/SinglyOwned.h:
```diff
@@ -0,0 +1,87 @@
+template<typename T>
+class SinglyOwned : private RefCountedWithInlineWeakPtr<T> {
+    friend class RefCountedWithInlineWeakPtr<T>;
+    template<typename U> friend SinglyOwnedPtr<U> adoptSinglyOwned(U*);
```
Maybe we can call this pointer type `OwnerPtr<U>`? Owner kind of implies singular ownership.
I think it would be nice to use a name that makes you think of `unique_ptr` or `UniqueRef` since the semantics are so similar.
How about `CanMakeUniqueInlineWeakPtr`?

geoffreygaren replied:
`OwnerPtr<U>` might sound like "pointer to owner" rather than "pointer that belongs to owner".

More Slack thread...

rniwa commented on Source/WTF/wtf/UniquelyOwned.h:
```diff
@@ -0,0 +1,87 @@
+    // CheckedPtr interface
+    void incrementCheckedPtrCount() const
+    {
+        this->ref();
```
Can't we just call `ref()` instead of calling `this->ref()`?
The same comment for other functions in this file.

geoffreygaren replied:
The private inheritance seems to take the names out of scope by default. Other option is `RefCountedWithInlineWeakPtr<T>::ref()`;

rniwa replied:
Doesn't this mean we'd delay the invocation of `~T` until the last `CheckedPtr` is destructed?
That sounds like a major semantics difference.
I don't think we want to change the timing of `~T` just because there is an outstanding `CheckedPtr<T>`.

rniwa commented on Source/WTF/wtf/RefCountedWithInlineWeakPtr.h:
```diff
@@ -0,0 +1,131 @@
+    void weakDerefSlowCase() const;
+} SWIFT_RETURNED_AS_UNRETAINED_BY_DEFAULT;
+
+template<typename T> void RefCountedWithInlineWeakPtr<T>::derefSlowCase() const
```
Do we need to split this into its own function? It seems like it's called exactly once in `deref()`?
I find it easier to follow if it was inlined there instead.

geoffreygaren replied:
I'm trying to help the optimizer by keeping the `++` and `--` inline and the rest out of line.

geoffreygaren replied:
(Subtracting the weak count, then checking for zero, then calling the destructor, then calling delete/free can add up to a lot of code since a lot of those functions are inline.)

rniwa commented on Source/WTF/wtf/RefCountedWithInlineWeakPtr.h:
```diff
@@ -0,0 +1,131 @@
+    weakDeref();
+}
+
+template<typename T> void RefCountedWithInlineWeakPtr<T>::weakDerefSlowCase() const
```
Ditto about merging this to `weakDeref`.

geoffreygaren replied:
Also for the optimizer.

rniwa commented on Source/WTF/wtf/UniquelyOwned.h:
```diff
@@ -0,0 +1,87 @@
+    void disown() const
+    {
+        this->refCountDebugger().willDelete();
+        this->deref();
```
I think we should be able to detect that this class takes T as an argument and eventually call delete static_cast<T*>(this) via this function in a superclass. That's probably better than hard-coding this class name.

rniwa commented on Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
```diff
@@ -6,6 +6,12 @@ Modules/push-api/PushManager.h
 Modules/storage/StorageManager.cpp
 Modules/webdatabase/DatabaseTask.h
 Modules/webdatabase/SQLTransactionBackend.h
+[ Mac ] inspector/InspectorFrontendHost.cpp
```
I don't understand why these entries are getting moved.
Can we undo these?
I'll look into why this is happening in the script.

geoffreygaren replied:
For reasons I didn't debug the script wasn't working for me, so I inserted entries manually and then sorted alphabetically.

geoffreygaren replied:
What's the intended sort order? Maybe I can fix it.

rniwa commented on Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
```diff
@@ -1,3 +1,6 @@
+<<<<<<< HEAD
```
Merge error.

rniwa commented on Source/WTF/wtf/InlineWeakPtr.h:
```diff
@@ -0,0 +1,189 @@
+namespace WTF {
+
+template<typename T>
+class InlineWeakPtr {
```
It's unfortunate that we need to create a brand new smart pointer type. This most certainly needs to be added to the static analyzer.

geoffreygaren replied:
Yeah, this is a bummer. I tried hard to make WeakPtr<T> work, but the tight coupling to many different flavors of WeakPtrImpl has gotten out of control, and I wasn't able to pull it apart.

rniwa commented on Source/WTF/wtf/RefCountedWithInlineWeakPtr.h:
```diff
@@ -0,0 +1,131 @@
+
+template<typename T> void RefCountedWithInlineWeakPtr<T>::derefSlowCase() const
+{
+    const_cast<T*>(static_cast<const T*>(this))->~T();
```
Wait, doesn't this mean we're using `this` after we had called this class's destructor already?
Is that a well defined behavior?

geoffreygaren committed:
Introduce UniquelyOwned and InlineWeakPtr and use them in RenderLayer
https://bugs.webkit.org/show_bug.cgi?id=304082
<rdar://problem/166403158>

Reviewed by Ryosuke Niwa.

Example usage:

class T : public UniquelyOwned<T> {
public:
    UniquelyOwnedPtr<T> create() { return adoptUniquelyOwned(new T); }

private:
    T() { }
};

class Owner {
    UniquelyOwnedPtr<T> m_t;
};

class X {
    // Data members must be weak
    InlineWeakPtr<T> m_t1;
    InlineWeakRef<T> m_t2;
};

void X::f()
{
    // Locals must be Checked
    if (CheckedPtr checkedT1 = m_t1)
        checkedT1->g();
    ChekckedRef { m_t2 }->g();

    // CheckedPtrs in the heap are also OK, but only if they're temporary,
    // and out of scope before ~T().
    Vector<CheckedPtr<T>> tmpVector;
    HashSet<CheckedPtr<T>> tmpHashSet;
}

This patch is neutral on Speedometer, MotionMark, and Membuster.

Chris tried diligently to adopt existing smart pointers in RenderLayer data
members, but it didn't work out:

    * SingleThreadWeakPtr is a performance regression

    * CheckedPtr crashes because our pointers outlive their pointees

    * Clearing pointers eagerly to prevent CheckedPtr crashes is a performance
    regression

Solution: InlineWeakPtr. InlineWeakPtr is a counter directly inside the object.

InlineWeakPtr is always faster than SingleThreadWeakPtr because it doesn't
require allocation or double-indirection.

InlineWeakPtr is sometimes smaller and sometimes larger than
SingleThreadWeakPtr:
    * It's 24 bytes smaller when the pointee is live
    * It's equal in size when the pointee and all pointers are dead
    * It's sizeof(T) - 16 bytes larger when the pointee is dead but a
    pointer is still live.

So InlineWeakPtr is always faster, and requires some thoughtfulness about
pointee lifetime and size.

InlineWeakPtr depends internally on reference counting
(RefCountedWithInlineWeakPtr), but we want to maintain the clarity that
RenderLayer is a single-owner object. So this patch also introduces UniquelyOwned
and UniquelyOwnedPtr, which indicate and enforce single ownership explicitly.

UniquelyOwned also improves the safety of CheckedPtr by eliminating
    * the window of time after scribbling and before crashing during which
    pointers borrowed before scribbling might still UAF
    * the whacky macro that must be included in all subclasses

I designed UniquelyOwned and RefCountedWithInlineWeakPtr to be useable in many
more places, including thread-safe places -- but this patch doesn't take those
steps yet.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/InlineWeakPtr.h: InlineWeakPtr is basically a copy of
RefPtr<T>, except that it calls weakRef() / weakDeref() instead of ref() / deref
(), and it checks for a non-zero refCount() before returning T*.

* Source/WTF/wtf/InlineWeakRef.h: Just like WeakRef<T>, for use in cases where
we want to verify memory safety but writing code to handle null would be absurd.

* Source/WTF/wtf/RefCountedWithInlineWeakPtr.h: The key insight of
RefCountedWithInlineWeakPtr is that the refcount gives us total control over
destruction, so we get a much tighter interface for checking validity of weak
pointers and CheckedPtrs.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/RefCntblBaseVirtualDtorExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
Updated expectations to copy RenderLayer CheckedPtr skips to RefCounted skips,
since UniquelyOwned is internally backed by refcounting.

* Source/WebCore/rendering/RenderLayer.h: Replace CanMakeSingleThreadWeakPtr<T>,
CanMakeCheckedPtr<T>, and WTF_OVERRIDE_DELETE_FOR_CHECKED_PTR with
UniquelyOwned<T>. Yay!

* Source/WebCore/rendering/RenderMarquee.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceContainer.cpp:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceContainer.h:
Adopt InlineWeakPtr now that RenderLayer requires it.

* Tools/TestWebKitAPI/Tests/WTF/CheckedPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/HashMap.cpp:
* Tools/TestWebKitAPI/Tests/WTF/HashSet.cpp:
* Tools/TestWebKitAPI/Tests/WTF/ListHashSet.cpp: Some API tests for new
CheckedPtr and InlineWeakPtr usage.

Canonical link: https://commits.webkit.org/304729@main
